Restructure Security chapter (#1359)
vaadin/docs: articles/flow/security/advanced-topics/securing-plain-java-app.adoc

---
title: Securing Plain Java Applications
description: Describes how to use the built-in security helpers in a plain Java application.
order: 50
layout: page
---

= Securing Plain Java Applications

As described in the <<{articles}/flow/security/enabling-security#, enabling security>> chapter, Vaadin Flow comes with built-in security helpers that are most convenient to implement with Spring Boot and Spring Security.

> While recommended, using Spring Security is not mandatory, however.

Reviewer: ```suggestion
Although it is recommended to use Spring Security, it is not mandatory.
```

.Separate instructions are available for Spring Boot and Spring Security.
[NOTE]
====
If yours is a Spring Boot project, then it is highly recommended to follow the instructions provided in the <<{articles}/flow/security/enabling-security#, enabling security>> chapter.
====

Vaadin built-in helpers enable a view-based access control mechanism that uses the [annotationname]`@AnonymousAllowed`, [annotationname]`@PermitAll`, [annotationname]`@RolesAllowed`, and [annotationname]`@DenyAll` annotations on view classes to define the access control rules.
**By default, any view without an annotation is secured; it acts as if it had the [annotationname]`@DenyAll` annotation.**

To enable and use this mechanism, the following should be added to a plain-Java Flow application (if it does not exist already):

- A security realm.
- A log-in view.
- Log-out capability.
- A [interfacename]`VaadinServiceInitListener` that adds [classname]`ViewAccessChecker` as the [classname]`BeforeEnterListener` of all UIs.
- A Service Provider under `META-INF/services` to load above the custom [classname]`ServiceInitListener` via the Java SPI loading mechanism.

== Configuring a Security Realm

> A _security realm_ is a mechanism for storing and mapping the users with their passwords and roles.

Reviewer: ```suggestion
A _security realm_ is a mechanism for storing and mapping users to their passwords and roles.
```

> However, the above class still needs to be loaded, so we recommend following the instructions in the next step.

Reviewer: ```suggestion
However, this class still needs to be loaded, so we recommend following the instructions in the next step.
```
